Update internal code for torch.lu_solve

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56613 Update internal code for torch.linalg.solve
* #56612 Update internal code for at::_lu_with_info
* **#56611 Update internal code for torch.lu_solve**

The goal of this refactoring is to make the `torch.linalg.solve`
to be a composition of calls to `lu_stub` and `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Replaced lu_solve_helper with DECLARE_DISPATCH for lu_solve_stub.
Removed unnecessary copy improving the performance (see https://github.com/pytorch/pytorch/pull/56611#issuecomment-824303673).
Split MAGMA-based `apply_lu_solve` into `apply_lu_solve_looped_magma`
and `apply_lu_solve_batched_magma`. This simplifies future dispatch to
cuSOLVER and cuBLAS.

Differential Revision: [D28142279](https://our.internmc.facebook.com/intern/diff/D28142279)

IvanYashchuk committed:
Update internal code for torch.lu_solve

The goal of this refactoring is to make the `torch.linalg.solve`
to be a composition of calls to `lu_stub` and `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Replaced lu_solve_helper with DECLARE_DISPATCH for lu_solve_stub.
Removed unnecessary copy improving performance.
Split MAGMA-based `apply_lu_solve` into `apply_lu_solve_looped_magma`
and `apply_lu_solve_batched_magma`. This simplifies future dispatch to
cuSOLVER and cuBLAS.

[ghstack-poisoned]

IvanYashchuk commented:
@heitorschueroff, if you have time could you please review and help to land this stack of PRs?

IvanYashchuk commented:
I removed one unnecessary copy improving the performance (`LU_data_broadcasted` in code).
Results for CPU:
```
|                          | before this PR | after this PR |
|--------------------------|----------------|---------------|
| torch.Size([2, 2])       | 5.5            | 3.7           |
| torch.Size([8, 8])       | 5.6            | 4.1           |
| torch.Size([16, 16])     | 5.8            | 4.2           |
| torch.Size([32, 32])     | 6.5            | 4.6           |
| torch.Size([64, 64])     | 8.0            | 5.7           |
| torch.Size([128, 128])   | 13.1           | 8.9           |
| torch.Size([256, 256])   | 32.0           | 18.5          |
| torch.Size([512, 512])   | 111.2          | 82.3          |
| torch.Size([1024, 1024]) | 1426.6         | 353.1         |
| torch.Size([2048, 2048]) | 13346.5        | 2021.1        |
| torch.Size([4096, 4096]) | 44818.3        | 7498.3        |

Times are in microseconds (us).
```

Results for CUDA:
```
|                          | before this PR | after this PR |
|--------------------------|----------------|---------------|
| torch.Size([2, 2])       | 1.3            | 1.3           |
| torch.Size([8, 8])       | 1.3            | 1.3           |
| torch.Size([16, 16])     | 1.2            | 1.3           |
| torch.Size([32, 32])     | 1.3            | 1.3           |
| torch.Size([64, 64])     | 1.4            | 1.3           |
| torch.Size([128, 128])   | 1.5            | 1.4           |
| torch.Size([256, 256])   | 1.6            | 1.3           |
| torch.Size([512, 512])   | 1.6            | 1.4           |
| torch.Size([1024, 1024]) | 1.8            | 1.6           |
| torch.Size([2048, 2048]) | 2.2            | 1.8           |
| torch.Size([4096, 4096]) | 2.8            | 2.2           |
| torch.Size([8192, 8192]) | 5.2            | 3.2           |

Times are in milliseconds (ms).
```

mruberry commented:
Awesome speedup, @IvanYashchuk!

@nikitaved would you take a look at this PR since you're working on lu?

IvanYashchuk committed:
Update on "Update internal code for torch.lu_solve"


The goal of this refactoring is to make the `torch.linalg.solve`
to be a composition of calls to `lu_stub` and `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Replaced lu_solve_helper with DECLARE_DISPATCH for lu_solve_stub.
Removed unnecessary copy improving the performance (see https://github.com/pytorch/pytorch/pull/56611#issuecomment-824303673).
Split MAGMA-based `apply_lu_solve` into `apply_lu_solve_looped_magma`
and `apply_lu_solve_batched_magma`. This simplifies future dispatch to
cuSOLVER and cuBLAS.

[ghstack-poisoned]

nikitaved commented:
@IvanYashchuk , could you please ping me once it is ready for review?

IvanYashchuk commented:
@nikitaved, this stack is ready for review

nikitaved commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3504,7 +3452,18 @@ Tensor lu_solve(const Tensor& self, const Tensor& LU_data, const Tensor& LU_pivo
-  return at::_lu_solve_helper(self_broadcasted, LU_data_broadcasted, LU_pivots_broadcasted);
+
+  // lu_solve_stub (apply_lu_solve) requires batched column major (Fortran-contiguous) tensors
+  // 'result' tensor is modified in-place and must be a copy of 'self_broadcasted'
+  Tensor result = cloneBatchedColumnMajor(self_broadcasted);
+
```
@IvanYashchuk , could you probably use `copyBatchedColumnMajor` instead of `cloneBatchedColumnMajor`. In my  experiments with `torch.linalg.lstsq` I got about 10-15% speed-up.

IvanYashchuk replied:
I tried and `cloneBatchedColumnMajor` was faster.

We need to materialize all matrices for batched input anyway. To keep the scope of the PR small I did not introduce new "features", that's better to be done in separate PRs.

nikitaved replied:
Agree, but we can avoid full matrix materialization for the CPU. But it indeed could be done in a separate PR.

nikitaved commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -531,6 +531,63 @@ void triangular_solve_kernel(Tensor& A, Tensor& B, Tensor& infos, bool upper, bo
+  * `b` -  [in] the right hand side matrix B
+           [out] the solution matrix X
+  * `lu` - [in] the LU factorization of matrix A (see at::_lu_with_info)
+  * `pivots` - [in] the pivot indices (see at::_lu_with_info)
```
nit: pivots returned by `_lu_with_info` encode the sequence of transpositions as written in the doc for `torch.lu`. I have to say the name `pivots` is a bit confusing in general, and lots of people had issues with understanding what `pivots` are until the most recent update of `torch.lu`'s doc.

nikitaved commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -2566,95 +2566,149 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda(const Tensor& self, bool som
-    scalar_t** lu_array;
-    scalar_t** b_array;
+  int info = 0;
+  for (decltype(batch_size) i = 0; i < batch_size; i++) {
+    scalar_t* b_working_ptr = &b_data[i * b_stride];
+    scalar_t* lu_working_ptr = &lu_data[i * lu_stride];
+    int* pivots_working_ptr = &pivots_data[i * pivots_stride];
```
Could we probably use `batch_iterator_with_broadcasting` here that was developed for `torch.linalg.lstsq`? Or is it something that is planned for a follow-up PR?

IvanYashchuk replied:
Definitely could be used here. I'd like to keep for a follow-up PR doing that simultaneously for all solve-like functions as it requires "before/after" benchmarking.

nikitaved commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -2566,95 +2566,149 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda(const Tensor& self, bool som
-    return at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+static void lu_solve_magma(const Tensor& b, const Tensor& lu, const Tensor& pivots) {
+  // TODO: compare performance and use the best performing option based on lu's sizes
+  if (b.dim() == 2) {
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(b.scalar_type(), "lu_solve_magma", [&]{
+      apply_lu_solve_looped_magma<scalar_t>(b, lu, pivots);
+    });
+  } else {
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(b.scalar_type(), "lu_solve_magma", [&]{
+      apply_lu_solve_batched_magma<scalar_t>(b, lu, pivots);
+    });
```
The internal comments on batched in looped versions mention that the batched version is faster than the looped for small input, i.e.
```
/*
  Solves the matrix equation A X = B
  X and B are n-by-nrhs matrices, A is represented using the LU factorization.
  This is an in-place routine, content of `b` is overwritten.
  This is a specialized batched variant, it is expected to be faster than the "looped" version only for small inputs.
```
In this code, however, the batched version is always chosen for a batched input of arbitrary size.

IvanYashchuk replied:
I kept the original conditions for choosing the single input variant over the batched one. A better condition that is based on the size of the input will be added separately in the future.
cc: @v0dro

nikitaved commented:
Looks great overall! I have only a couple of comments/suggestions.

nikitaved commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -531,6 +531,63 @@ void triangular_solve_kernel(Tensor& A, Tensor& B, Tensor& infos, bool upper, bo
+  auto leading_dimension = std::max<int64_t>(1, n);
+
+  int info = 0;
+  for (const auto i : c10::irange(batch_size)) {
+    scalar_t* b_working_ptr = &b_data[i * b_stride];
+    scalar_t* lu_working_ptr = &lu_data[i * lu_stride];
+    int* pivots_working_ptr = &pivots_data[i * pivots_stride];
```
Could we use `batched_iterator_with broadcasting` here? This is to avoid memory inefficiencies and also simplify the code a bit...

IvanYashchuk replied:
I was hesitant to use it because it requires further benchmarking. We are going to add cuSOLVER backend for this function soon which launches kernels asynchronously and I'm not sure how would `batched_iterator_with broadcasting` behave in this situation.
And the specialized batched variants require full materialization of batched inputs.

nikitaved replied:
Got it, but this is the CPU code, so we do not depend on cuSOLVER here I presume?

nikitaved reviewed (approved):
LGTM. Some modifications, such as the usage of `copyBatchedColumnMajor` instead of `cloneBatchedColumnMajor`, and memory improvements with `batch_iterator_with_broadcasting` for the CPU/Magma (i.e. batch loop-based paths) could be done in a follow-up PR.

mruberry reviewed (approved):
Thanks @IvanYashchuk and @nikitaved!

@IvanYashchuk ping me when you'd like this merged

IvanYashchuk committed:
Update on "Update internal code for torch.lu_solve"


The goal of this refactoring is to make the `torch.linalg.solve`
to be a composition of calls to `lu_stub` and `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Replaced lu_solve_helper with DECLARE_DISPATCH for lu_solve_stub.
Removed unnecessary copy improving the performance (see https://github.com/pytorch/pytorch/pull/56611#issuecomment-824303673).
Split MAGMA-based `apply_lu_solve` into `apply_lu_solve_looped_magma`
and `apply_lu_solve_batched_magma`. This simplifies future dispatch to
cuSOLVER and cuBLAS.

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, I rebased and resolved conflicts. Let's merge this PR, the next PR in the stack awaits @nikitaved's approval. And I still need to resolve the Windows failure of the last PR in the stack (https://github.com/pytorch/pytorch/pull/56613).

mruberry commented:
Incredibly our internal tools are still complaining about this one. Sorry @IvanYashchuk. If you can rebase this I'll get this landed ASAP

IvanYashchuk committed:
Update on "Update internal code for torch.lu_solve"


The goal of this refactoring is to make the `torch.linalg.solve`
to be a composition of calls to `lu_stub` and `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Replaced lu_solve_helper with DECLARE_DISPATCH for lu_solve_stub.
Removed unnecessary copy improving the performance (see https://github.com/pytorch/pytorch/pull/56611#issuecomment-824303673).
Split MAGMA-based `apply_lu_solve` into `apply_lu_solve_looped_magma`
and `apply_lu_solve_batched_magma`. This simplifies future dispatch to
cuSOLVER and cuBLAS.

Differential Revision: [D28142279](https://our.internmc.facebook.com/intern/diff/D28142279)

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, I rebased this PR.

heitorschueroff commented:
> @heitorschueroff, if you have time could you please review and help to land this stack of PRs?

@IvanYashchuk Sorry I missed this comment, but it looks like the PR is ready to land now. @mruberry let me know if you want me to land this.

IvanYashchuk committed:
Update on "Update internal code for torch.lu_solve"


The goal of this refactoring is to make the `torch.linalg.solve`
to be a composition of calls to `lu_stub` and `lu_solve_stub`.
Once `lu_stub` and `lu_solve_stub` have cuSOLVER-based codepath,
`torch.linalg.solve` will have it as well.

Replaced lu_solve_helper with DECLARE_DISPATCH for lu_solve_stub.
Removed unnecessary copy improving the performance (see https://github.com/pytorch/pytorch/pull/56611#issuecomment-824303673).
Split MAGMA-based `apply_lu_solve` into `apply_lu_solve_looped_magma`
and `apply_lu_solve_batched_magma`. This simplifies future dispatch to
cuSOLVER and cuBLAS.

Differential Revision: [D28142279](https://our.internmc.facebook.com/intern/diff/D28142279)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28142279).

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28142279).
